### PR TITLE
RzIL: add `aezsue` to support `Step until PC equals given address and output VM changes`

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/arch/il/analysis_il.c [hunk 1]
```diff
@@ -251,20 +253,37 @@ RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(RZ_NONNULL RzAnalysis
 	}
 
 	RzAnalysisILStepResult res = RZ_ANALYSIS_IL_STEP_RESULT_SUCCESS;
+	ut64 addr = 0;
 	while (cond(vm, user)) {
-		ut64 addr = rz_bv_to_ut64(vm->vm->pc);
```
Comment (p371jah) [inline librz/arch/il/analysis_il.c:255]: Why?

File: librz/arch/il/analysis_il.c [hunk 2]
```diff
@@ -228,17 +228,19 @@ RZ_API bool rz_analysis_il_vm_sync_to_reg(RzAnalysisILVM *vm, RZ_NONNULL RzReg *
...
- * Then it repeatedly disassembles an instruction at the program counter of the vm and executes it as long as cond() returns true.
- * Finally the contents are optionally synced back to \p reg.
- *
- * \return and indicator for which error occured, if any, or RZ_ANALYSIS_IL_STEP_RESULT_SUCCESS if cond() returned false
- */
-RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL RzAnalysisILVM *vm, RZ_NULLABLE RzReg *reg,
-	bool (*cond)(RzAnalysisILVM *vm, void *user), void *user) {
+static void il_events(RzILVM *vm, RzStrBuf *sb) {
+	void **it;
+	RzILEvent *evt;
```
Comment (p371jah) [inline librz/arch/il/analysis_il.c]: No reason to define it outside the `foreach` scope

File: librz/il/il_vm_eval.c [hunk 3]
```diff
@@ -261,7 +261,12 @@ RZ_API void rz_il_vm_mem_storew(RzILVM *vm, RzILMemIndex index, RzBitVector *key
 		return;
 	}
 	RzBitVector *old_value = rz_il_mem_loadw(mem, key, rz_bv_len(value), vm->big_endian);
-	rz_il_mem_storew(mem, key, value, vm->big_endian);
+	if (!rz_il_mem_storew(mem, key, value, vm->big_endian)) {
+		RZ_LOG_ERROR("StoreW mem %u 0x%llx failed\n", (unsigned int)index, rz_bv_to_ut64(key));
```
Comment (p371jah) [inline librz/il/il_vm_eval.c:265]: I guess `old_value` leaks here

File: librz/il/il_vm_eval.c [hunk 4]
```diff
@@ -261,7 +261,12 @@ RZ_API void rz_il_vm_mem_storew(RzILVM *vm, RzILMemIndex index, RzBitVector *key
...
+		return;
+	}
+	RzBitVector *new_value = rz_il_mem_loadw(mem, key, rz_bv_len(value), vm->big_endian);
+	assert(rz_bv_eq(new_value, value));
```
Comment (p371jah) [inline librz/il/il_vm_eval.c]: This is just a debug code?

File: librz/arch/il/analysis_il.c [hunk 5]
```diff
@@ -228,17 +228,19 @@ RZ_API bool rz_analysis_il_vm_sync_to_reg(RzAnalysisILVM *vm, RZ_NONNULL RzReg *
...
+	rz_pvector_foreach (vm->events, it) {
+		evt = *it;
+		rz_il_event_stringify(evt, sb);
+		rz_strbuf_append(sb, "\n");
+	}
+}
+
+static RzAnalysisILStepResult analysis_il_vm_step_while(
+	RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL RzAnalysisILVM *vm, RZ_NULLABLE RzReg *reg,
+	bool with_events, bool (*cond)(RzAnalysisILVM *vm, void *user), void *user) {
```
Comment (p371jah) [inline librz/arch/il/analysis_il.c]: `cond` is `RZ_NONNULL` and `user` is `RZ_NULLABLE`

File: librz/arch/il/analysis_il.c [hunk 6]
```diff
@@ -251,20 +253,37 @@ RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(RZ_NONNULL RzAnalysis
...
+				il_events(vm->vm, &sb);
+
+				char *il_stmt = rz_strbuf_get(&sb);
+				rz_cons_printf("0x%llx [%x%x%x%x] %s\n"
+					       "0x%llx %s\n",
+					addr, code[0], code[1], code[2], code[3],
+					op.mnemonic, addr, il_stmt);
+				rz_strbuf_fini(&sb);
+			}
+
```
Comment (p371jah) [inline librz/arch/il/analysis_il.c]: Invert the condition

**Review by p371jah — CHANGES_REQUESTED** [review]

File: librz/arch/il/analysis_il.c (hunk 5 above)
Comment (wargio) [reply librz/arch/il/analysis_il.c]: maybe declare as `typedef` the `bool (*cond)(RzAnalysisILVM *vm, void *user)` function

File: librz/arch/il/analysis_il.c (hunk 6 above)
Comment (b1llow) [reply librz/arch/il/analysis_il.c]: It seems to me that the “Invert the condition” here is not appropriate.

File: librz/arch/il/analysis_il.c [hunk 7]
```diff
@@ -280,6 +299,53 @@ RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(RZ_NONNULL RzAnalysis
...
+ *
+ * \return RZ_ANALYSIS_IL_STEP_RESULT: Enumeration value indicating the outcome of the stepping operation.
+ *         Possible values (implementation specific):
+ *             - RZ_ANALYSIS_IL_STEP_OK: Successful execution of the while loop step.
+ *             - RZ_ANALYSIS_IL_STEP_ERROR: Encountered an error during execution.
+ *             - RZ_ANALYSIS_IL_STEP_INVALID: Invalid arguments or state resulted in undefined behavior.
+ */
+RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(
+	RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL RzAnalysisILVM *vm, RZ_NULLABLE RzReg *reg,
+	RZ_NONNULL cond_callback cond, RZ_NULLABLE void *user) {
```
Comment (wargio) [inline librz/arch/il/analysis_il.c]: If `cond_callback` is used in the `RZ_API` then is worth moving it with a proper name into the header.

File: librz/arch/il/analysis_il.c (hunk 6 above)
Comment (p371jah) [reply librz/arch/il/analysis_il.c]:
```c
if (!ilop) {
	res = RZ_ANALYSIS_IL_STEP_INVALID_OP;
	rz_analysis_op_fini(&op);
	break;
}
bool succ = rz_il_vm_step(vm->vm, ilop, addr + (op.size > 0 ? op.size : 1));
if (!succ) {
	res = RZ_ANALYSIS_IL_STEP_IL_RUNTIME_ERROR;
	rz_analysis_op_fini(&op);
	break;
}
```

Comment (p371jah) [reply librz/arch/il/analysis_il.c]: `op` could be moved outside the loop body to reduce `rz_analysis_op_fini(&op);` lines count

File: librz/arch/il/analysis_il.c [hunk 8]
```diff
@@ -250,36 +251,96 @@ RZ_API RzAnalysisILStepResult rz_analysis_il_vm_step_while(RZ_NONNULL RzAnalysis
...
 	}
 
+	RzAnalysisOp op = { 0 };
 	RzAnalysisILStepResult res = RZ_ANALYSIS_IL_STEP_RESULT_SUCCESS;
 	while (cond(vm, user)) {
 		ut64 addr = rz_bv_to_ut64(vm->vm->pc);
 		ut8 code[32] = { 0 };
 		analysis->read_at(analysis, addr, code, sizeof(code));
-		RzAnalysisOp op = { 0 };
 		rz_analysis_op_init(&op);
```
Comment (p371jah) [inline librz/arch/il/analysis_il.c:260]: Move it outside

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
LGTM once @pelijah feedback is addressed.

File: librz/arch/il/analysis_il.c (hunk 8 above)
Comment (b1llow) [reply librz/arch/il/analysis_il.c:260]: I don't get it. It seems to me that `rz_analysis_op_init` needs to be called before every call to `rz_analysis_op` and `rz_analysis_op_fini` needs to be called after every call to `rz_analysis_op`. 

So ‘Move it outside’ doesn't work? @pelijah

Comment (p371jah) [reply librz/arch/il/analysis_il.c:260]: Weird, `rz_analysis_op_fini()` doesn't work as expected so leave it as is.

Comment (p371jah) [reply librz/arch/il/analysis_il.c:260]: @XVilka We definetely should standardize `*_fini()` APIs behaviour.

**Review by p371jah — APPROVED** [review]

**Pull request merged**